src/algebraic_geometry/EllipticCurve/weierstrass.lean

```diff
@@ -0,0 +1,287 @@
+/-
+Copyright (c) 2021 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard, David Kurniadi Angdinata
+-/
+
+import algebra.cubic_discriminant
+import tactic.linear_combination
+
+/-!
+# Weierstrass equations of elliptic curves
+
+We give a working definition of an elliptic curve as a non-singular Weierstrass curve given by a
+Weierstrass equation, which is mathematically accurate in many cases but also good for computation.
+
+## Mathematical background
+
+Let `S` be a scheme. The actual category of elliptic curves over `S` is a large category, whose
+objects are schemes `E` equipped with a map `E → S`, a section `S → E`, and some axioms (the map
+is smooth and proper and the fibres are geometrically-connected one-dimensional group varieties). In
+the special case where `S` is the spectrum of some commutative ring `R` whose Picard group is zero
+(this includes all fields, all PIDs, and many other commutative rings) it can be shown (using a lot
+of algebro-geometric machinery) that every elliptic curve `E` is a projective plane cubic isomorphic
+to a Weierstrass curve given by the equation $y^2 + a_1xy + a_3y = x^3 + a_2x^2 + a_4x + a_6$ for
+some $a_i$ in `R`, and such that a certain quantity called the discriminant of `E` is a unit in `R`.
+If `R` is a field, this quantity divides the discriminant of a cubic polynomial whose roots over a
+splitting field of `R` are precisely the x-coordinates of the non-zero 2-torsion points of `E`.
+
+## Main definitions
+
+ * `weierstrass_curve`: a Weierstrass curve over a commutative ring.
+ * `weierstrass_curve.variable_change`: the Weierstrass curve induced by a change of variables.
+ * `weierstrass_curve.base_change`: the Weierstrass curve base changed over an algebra.
+ * `weierstrass_curve.two_torsion_polynomial`: the 2-torsion polynomial of a Weierstrass curve.
+ * `elliptic_curve`: an elliptic curve over a commutative ring.
```

Comment:
I think `EllipticCurve` is capitalized because @kbuzzard wanted to make it a category, which hasn't been done. Does he still want the category based on Weierstrass equation? If so, is it meaningful to consider the category `WeierstrassCurve`? If not, then the name change to `elliptic_curve` is good, and we should probably also change the name of the directory.

Reply:
Should we wait for @kbuzzard's response? Or maybe you've talked to him. Let me request his review and we should be good to go given his approval.

Reply:
I've asked him privately and he said (paraphrased) he doesn't mind, he originally just wanted some definition on mathlib.